### PR TITLE
[WIP] Database backup file monitoring

### PR DESCRIPTION
**Pull request by medyakovvit**
Added ability to specify folder for database backup.

### CODE REVIEW

File: src/MainWindow.cpp [hunk 1]
```diff
@@ -101,6 +101,13 @@ MainWindow::MainWindow(WSClient *client, QWidget *parent) :
     connect(wsClient, &WSClient::wsDisconnected, this, &MainWindow::updatePage);
     connect(wsClient, &WSClient::connectedChanged, this, &MainWindow::updatePage);
     connect(wsClient, &WSClient::statusChanged, this, &MainWindow::updatePage);
+    connect(wsClient, &WSClient::databaseBackupFolderChanged, [=](QString backupFolder){
+        // use some default folder
+        if(backupFolder.isEmpty())
+            backupFolder = QCoreApplication::applicationDirPath();
```
[the remark this answers is not in the document]
Comment (medyakovvit) [reply src/MainWindow.cpp]: As decided, _backupFolder_ should be changed to _backupFile_. So, what default file name should be used here?

**Comment by limpkin** [comment]
nothing.

File: src/MainWindow.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (raoulh) [reply src/MainWindow.cpp]: I would say it it's empty, the backup is not enabled, right?

**Comment by limpkin** [comment]
yup.

**Comment by raoulh** [comment]
I forgot also to say that you added a new Websocket API, you need to add it also to the debug web page (for documenting the API) here: https://github.com/raoulh/moolticute/blob/master/data/debug/app/scripts/main.js and rebuild the webapp

**Comment by limpkin** [comment]
@medyakovvit just FYI part of my email will be about this.

**Comment by medyakovvit** [comment]
Hi.
I added code to check change numbers. I added widget with question asking user for updating with more recent version of credentials (from backup file or device). Clicking "Yes" just will show stub message for user. I just want to confirm that i'm going in right way.

**Comment by limpkin** [comment]
Hey there!

Thanks a lot for that!
I'll try to test your code tonight :)

**Comment by medyakovvit** [comment]
Ok, great.

Lets consider situation: change number in the device - 6, in backup file - 7. We asking user for updating and user chooses "No" or ignores question. And then adds one new sevice/login/password. Change number in the device now  - 7. And there will be no mismatch after next check. Is it make sense?

**Comment by limpkin** [comment]
That's a valid point... And a very nice find.
I guess I would display another warning saying that by denying this prompt
the user may lose changes.
After... It's the user's responsibility after all.

**Comment by limpkin** [comment]
Hey there!

So I finally managed to test your PR:
- when selecting a backup file, I'm getting a "save as" dialog when it should be a "read" one. 
- as a consequence, it is prompting me to overwrite a file I want to **monitor**
- the (very very nice) notification "credentials on the device are more recent..." should be displayed just after a new credential has been added to the device, either through the app or through the plugin. I'm guessing you get a notification when the change number is updated. You'll therefore need to trigger getChangeNumbers() when an addition to the DB is made (through setCredential() for example)

Except that, your code is working beautifully. I'm really impressed.

**Comment by limpkin** [comment]
also, I just thought that it may be a great idea to display a big red warning side inside the tray icon to notify the user for action :)

**Comment by medyakovvit** [comment]
Hi.
I added checking for change numbers after updating credentials through Moolticute.

**Comment by limpkin** [comment]
Wonderful job, thanks a lot!

-  I only get the notification message if the daemon is started after the app... in some cases (linux) the daemon is started before the app
- It may be a wonderful idea when getting the "credentials on the device are more recent..." to directly overwrite the file being monitored and not ask the user to select a file again. When the export is done, maybe display a small notification somewhere saying something like "monitored backup file updated"
- Maybe also trigger your changeNumber check when the user is importing another DB... for some (stupid) reason the user may import an old DB
- It may also be a great idea when getting the "credentials in the backup file are more recent..." to directly select the file being monitored and not ask the user to select the file being monitored. When the import is done, maybe display a small notification somewhere saying something like "mooltipass database updated"

**Comment by medyakovvit** [comment]
Hi, pushed changes.
- changing monitoring file will call checking for credentials change numbers;
- monitoring file is used for import/export by default, If monitoring file not specified user is asked for choosing file and this file will be used as monitoring file.

> in some cases (linux) the daemon is started before the app
wow, How that possible if daemon is launched from app with QProcess? Will look at.

**Comment by limpkin** [comment]
hey there!
thanks for the changes! I should be able to check tomorrow.
About linux: simple: moolticuted launched using systemd :)

**Comment by limpkin** [comment]
A few other errors: 

1) 
image

2) 
- use the "change..." button to setup a backup file to monitor
- clicking the import from file button then doesn't allow the user to select another file to import (the user may need to do that)

3) when db file monitoring isn't setup, and the user import a file, this file will be setup as being monitored: please disable that

4) we may need an additional button ("delete") to remove the database file being monitored

5) can't reproduce it reliably: sometimes, when the file monitoring solution seems to be "disabled" and importing another database file, the "credentials on teh device are more recent" message appears **even before the user accepted the import prompt on the mini**

6) it seems there's no proper monitoring in the backend: by changing the monitored file with a more recent one, i don't get the prompt on moolticute, **only by restarting the daemon** i get the correct message

**Comment by medyakovvit** [comment]
> we may need an additional button ("delete") to remove the database file being monitored

Do you want just remove file or also clean the value in "Database file:" field?

**Comment by limpkin** [comment]
just disable the monitoring feature: clean the value only

**Comment by medyakovvit** [comment]
I pushed some changes.
Try to check 5) and 6) again.

**Comment by limpkin** [comment]
1) solved
2) solved
3) solved
4) done
5) seems ok
6) **not solved**

new problems:
7) when setting a new file to be monitored, which is more recent than the DB on the device and clicking yes on the "credentials... are more recent..." prompt, the user is asked to select an import file instead of automatically using the monitored file

wishes:
8) instead of a notification inside moolticute main window when the monitored file is more recent, have a new window (but only for that use case)

**Comment by raoulh** [comment]
6. A QFileSystemWatcher has to be used for monitoring the file.
http://doc.qt.io/qt-5/qfilesystemwatcher.html

Also, there are conflicts to be resolved with master

**Comment by medyakovvit** [comment]
6. oh! you mean that you change file in folder and not monitored file's path through the app. Misunderstood that.

**Comment by raoulh** [comment]
6. Yes, the typical use case is ig you have the backup file that is on folder from a cloud drive (dropbox, gdrive,...) then the file update should be taken into account by moolticude

**Comment by medyakovvit** [comment]
Hi, i added:
- watching for backup file with QFileSystemWatcher;
- changes for 7) and 8);
- fixed conflicts;
- reading change number from new format of backup file (with encrypted data).

**Comment by limpkin** [comment]
Wow.. thanks for the great work! I'll have a look today :)

**Comment by limpkin** [comment]
6) **does not work** on windows at least

not sure if related but:

`DEBUG: WSClient.cpp:124 - New message:  {"data":2,"msg":"credentials_change_numbers"}
shell\comdlg32\fileopensave.cpp(14267)\comdlg32.dll!750E7C7D: (caller: 751168FC) ReturnHr(16) tid(3db0) 80004005 Unspecified error
    CallContext:[\PickerModalLoop] 
shell\comdlg32\fileopensave.cpp(14267)\comdlg32.dll!750E7C7D: (caller: 751168FC) ReturnHr(17) tid(3db0) 80004005 Unspecified error
    CallContext:[\PickerModalLoop]`

7) solved
8) done

new cases:
9) when importing either a newer or older database manually, there is no prompt to export or import the DB to the monitored file

**Comment by limpkin** [comment]
**big problem**

10) when clicking the export to file button, the file is automatically exported into the monitored file instead of asking the user where to store it!

**Comment by limpkin** [comment]
11) when exporting to a file, this file is selected as monitored file, **please remove that**

To simplify your testing, I'll describe here all the test cases that need to be checked for this PR to be approved

**Monitoring not setup**
1) clicking export to file successfully exports the data, no monitored file change
2) clicking import from file with said file works, no monitored file change
3) same as 1) with "generate mooltiapp compatible file" selected
4) same as 2) with "generate mooltiapp compatible file" selected

Now considering 2 DB backup files: old DB file **A**, new DB file **B**

**Monitoring setup on file A, with DB as file A**
1) clicking export to file exports the data **to a file that the user can pick**
2) clicking import from file, selecting file A works, no monitored file change nor any prompt
3) clicking import from file, selecting file B works, a message appears to export to monitored file
4) changing monitored file to file B creates a popup inviting to import data
5) set monitored file to file A, replacing through the filesystem the fileA with fileB creates a popup inviting to import data

**Monitoring setup on file B, with DB as file B**
1) clicking export to file exports the data **to a file that the user can pick**
2) clicking import from file, selecting file B works, no monitored file change nor any prompt
3) clicking import from file, selecting file A works, a popup inviting the user to import data from monitored file appears
4) changing monitored file to file A creates a message inviting to export the DB
5) set monitored file to file B, replacing through the filesystem the fileB with fileA creates a message inviting to export data

**+ same tests as above, selecting "generate mooltiapp compatible file"**

**Comment by medyakovvit** [comment]
Hi.
Tnak you for preparing of check list.
I need some clarification about MooltiApp compatible file.
As i understand such file should meet requirenments:
1. is JSON array,
2. 9th element = "mooltipass",
3. length of JSON array = 10.
Am i right?

**Comment by limpkin** [comment]
@medyakovvit that is correct... but FYI it's exactly the same as before. We just changed the "normal export file format" to an encrypted one and renamed the previous one to "mooltiapp compatible"

**Comment by medyakovvit** [comment]
So, (dataArray[9].toString() == "mooltipass" && dataArray.size() == 10) it's not needed here https://github.com/mooltipass/moolticute/blob/fd8e499b19b8641fb7960cba8ba7516ca6160aa8/src/MPDevice.cpp#L4926

**Comment by limpkin** [comment]
actually it is, as we can now deal with 2 kinds of backup files (mooltiapp compatible and the encrypted "moolticute" one)

**Comment by limpkin** [comment]
Well, it's the GUI interacting with the user though... Raoul, what do you
think?

**Comment by medyakovvit** [comment]
But user could be asked to choose file before starting exporting data.

**Comment by raoulh** [comment]
Hi,

Ok, I did a full review of your code. Unfortunately there is something in the concept that is wrong: The QFileSystemWatcher is done in the daemon. This is not how it need to work.

Here is the correct workflow:

 - The GUI needs to watch the file change
 - If file is changed on disk, it compares changeNumber from the device (the daemon needs to send the actual changeNumber from the device to the GUI)
 - If the changeNumber differs, GUI should ask the user, then use the existing `import_database` JSON call

The daemon should never be aware of any filename or path. This is done like that because the daemon could be running as a service with different privilege or even on an other computer.

You would need to move the QFileSystemWatcher code to the GUI.
Sorry we misunderstood each other. We need to be more clear in the future.

Thanks for all your work!

**Comment by limpkin** [comment]
@medyakovvit : do you have any question regarding @raoulh 's comment?

**Comment by medyakovvit** [comment]
I don't have questions for now.

**Comment by azubieta** [comment]
@raoulh Have you considered the case where the file was changed in both parts and have the same change number but they differ in content?
Maybe having an checsum or something alike would be useful in order to detect divergences more precisely.

Any way a merge operation will be required instead of a simple import.

**Comment by limpkin** [comment]
@azubieta we have warnings to prevent this kind of scenario but that may indeed happen.

**Comment by azubieta** [comment]
Working on merge the UI.

**Comment by limpkin** [comment]
Shouldn't the client simply send the filename to the back end, the daemon
do the monitoring and notify the client when an action is required?

**Comment by azubieta** [comment]
@limpkin  Please take a look at the @raoulh comment above where he explains the desired work flow. He states that file monitoring should be done at the client side and only call import/export when required.

**Comment by limpkin** [comment]
Oh yeah, that's right

**Comment by limpkin** [comment]
@medyakovvit thanks again for your great work that @azubieta was able to use. Please let me know in private if you'd like something else to contribute on :)

**Pull request closed without merging**